src/clj/puppetlabs/services/master/master_core.clj

```diff
@@ -430,6 +478,62 @@
                              jruby-service)
         (environment-not-found environment)))))
 
+(defn exception-matches?
+  [^Exception e pattern]
+  (->> e
+       .getMessage
+       (re-find pattern)
+       boolean))
+
+(schema/defn ^:always-validate
+  task-details :- TaskDetails
+  [jruby-service :- (schema/protocol jruby-protocol/JRubyPuppetService)
+   jruby-instance
+   environment-name :- schema/Str
+   module-name :- schema/Str
+   task-name :- schema/Str]
+  (-> (jruby-protocol/get-task-data jruby-service
+                                    jruby-instance
+                                    environment-name
+                                    module-name
+                                    task-name)
+      sort-nested-info-maps
+      (describe-task environment-name module-name)))
+
+(schema/defn ^:always-validate
+  task-details-fn :- IFn
+  "Middleware function for constructing a Ring response from an incoming
+  request for detailed task information."
+  [jruby-service :- (schema/protocol jruby-protocol/JRubyPuppetService)]
+  (fn [request]
+    (let [environment (jruby-request/get-environment-from-request request)
+          module (get-in request [:route-params :module-name])
+          task (get-in request [:route-params :task-name])]
+      (let [?task-details (try (task-details jruby-service
+                                             (:jruby-instance request)
+                                             environment
+                                             module
+                                             task)
+                            (catch RaiseException e
+                              (cond
+                                (exception-matches? e #"(EnvironmentNotFound)")
+                                ::missing-environment
+
+                                (exception-matches? e #"(MissingModule)")
+                                ::missing-module
+
+                                (exception-matches? e #"(TaskNotFound)")
+                                ::missing-task
+
+                                :else
+                                (throw e))))]
+        (case ?task-details
```

Comment:
Is there a reason not to inline these into the `cond` above and just have the `json-response` after the `task-details` call in your try?